### PR TITLE
add parser for quorum sets, fix quorum set equality

### PR DESCRIPTION
**Pull request by rjwalters**
Soundtrack of this PR: [bass](https://soundcloud.com/shadowtrixmusic/dropical-that-feel)

### Motivation

I'd like to be able to specify arbitrary network configurations as lists of quorum sets. This PR allows a node to specify a quorum set using a simple grammar, such "([3],1,2,3,4,([2],5, 6,([1],8,7)))" meaning that this node (perhaps node 0) needs 3 of 1,2,3,4, and an inner set to move fwd, etc.

This is also preparatory work for revisiting some of the quorum slice machinery that will let us test networks for quorum intersection.

### In this PR
* adds Ord trait to NodeID
* allows Quorum sets with differently ordered members to be equal
* adds a PEG-based grammar for quorum set expression using pest.rs

### Future Work
* start enumerating quorum slice sets to makes sure test networks have intersection

### CODE REVIEW

**Comment by eranrund** [comment]
I don't see why this is justified. Your example `([3],1,2,3,4,([2],5, 6,([1],8,7)))` can be written as this, which I also find sufficiently readable and requires no extra code:
```
                quorum_set = { threshold = 3, members = [
                    # Node 1
                    { type = "Node", args = "id1" },

                    # Node 2
                    { type = "Node", args = "id2" },

                    # Node 3
                    { type = "Node", args = "id3" },

                    # InnerSet
                    { type = "InnerSet", args = { threshold = 2, members = [
                        { type = "Node", args = "node5" },
                        { type = "Node", args = "node6" },
                        { type = "InnerSet", args = { threshold = 1, members = [
                            { type = "Node", args = "node8" },
                            { type = "Node", args = "node7" },
                         ] }
                    ] } },
                ] }

```

**Comment by eranrund** [comment]
If you feel strongly this is necessary for tests then I am willing to accept if if more people agree on adding it.

**Comment by rjwalters** [comment]
> If you feel strongly this is necessary for tests then I am willing to accept if if more people agree on adding it.

I prefer the shorter syntax. It's also longer/more complicated than your example because of the way `NodeID` is constructed. 

I want to provide lists of 10s of quorum sets as a way of defining arbitrary networks. I'm also planning to pull down stellar's topology from stellarbeat for comparison testing too -- it's going to be nice to have tooling for managing quorum sets as simple strings.

**Comment by eranrund** [comment]
My preference is to have the least amount of code to maintain. Why would my example need to be longer? This is pretty much taken from the tests in `consensus/service/src/config.rs`.

This can also be done programmatically without being particularly long either: https://github.com/mobilecoinofficial/mobilecoin/blob/master/consensus/scp/src/quorum_set.rs#L522

I remain unconvinced that we need a third way to construct QuorumSets.

File: consensus/scp/src/quorum_set.rs [hunk 1]
```diff
@@ -28,29 +51,22 @@ pub struct QuorumSet<ID: GenericNodeId = NodeID> {
...
-pub enum QuorumSetMember<ID: GenericNodeId> {
-    /// A single trusted entity with an identity.
-    Node(ID),
-
-    /// A quorum set can also be a member of a quorum set.
-    InnerSet(QuorumSet<ID>),
+impl<ID: GenericNodeId> PartialEq for QuorumSet<ID> {
+    fn eq(&self, other: &QuorumSet<ID>) -> bool {
+        if self.threshold == other.threshold && self.members.len() == other.members.len() {
+            // sort before comparing
```
Comment (sugargoat) [inline consensus/scp/src/quorum_set.rs:57]: This is useful - another option would be to make `Vec<>` into `HashSet<>` - but that would be a larger change.

**Comment by rjwalters** [comment]
I peeled off the quorum set equality fix in #235. I will rework the parsing stuff in a new PR.

**Pull request closed without merging**